scala/data_structures/bloom_filter.scala

```diff
@@ -1,66 +1,78 @@
 package data_structures
 
+import scala.util.hashing.MurmurHash3.stringHash
+import Math.{abs, ceil, log, pow, round}
+
 case class BloomFilter[A](
   nbOfItems: Int,
   falsePositiveProbability: Float,
   private val array: Array[Boolean],
-  private val hashFunctions: Array[String => Int]
+  private val numberOfHashFunctions: Int,
+  private val hashSeed: Int
 ) {
 
+  private val hashFunctions: Stream[String => Int] =
+    (1 to numberOfHashFunctions)
+      .toStream
+      .map(i => (str: String) => abs(stringHash(str, hashSeed + i)) % array.length)
+
   def +=(item: A): BloomFilter[A] = {
     val itemString = item.toString
-    val hashes = this.hashFunctions.map(_(itemString))
-    val updatedArray = hashes.foldLeft(this.array)((acc, hash) => acc.updated(hash, true))
-    BloomFilter(nbOfItems, falsePositiveProbability, updatedArray, hashFunctions)
+    val hashes = hashFunctions.map(_(itemString))
+    val updatedArray = hashes.foldLeft(array)((acc, hash) => acc.updated(hash, true))
+    this.copy(array=updatedArray)
   }
 
   def ++=(items: Seq[A]): BloomFilter[A] = items.foldLeft(this)(_ += _)
 
-  def contains(item: A): Boolean = {
+  def mayContain(item: A): Boolean = {
     val itemString = item.toString
-    this.hashFunctions.forall(fn => this.array(fn(itemString)))
+    hashFunctions.forall(fn => array(fn(itemString)))
+  }
+
+  def approxNumberOfItems(): Int = {
+    val setBits = array.filter(identity).length.toDouble
+    val totalBits = array.length.toDouble
+    round(
+      (-totalBits / numberOfHashFunctions.toDouble) * log((1D - (setBits / totalBits)))
+    ).toInt
```

Comment:
Thanks @jonathanlofgren for the inspiration

Reply:
Similar to this comment https://github.com/nugit/LYANL/pull/26#discussion_r276513062, would be great if there's a comment on what this is doing so it will be easier to revisit or for newcomers to know what is happening 👍

Reply:
We basically translated this https://en.wikipedia.org/wiki/Bloom_filter#Approximating_the_number_of_items_in_a_Bloom_filter into Scala and Rust :slightly_smiling_face: 
But yeah maybe we can add a comment with a link